src/configureWorkspace/scaffolding.ts

```diff
@@ -0,0 +1,144 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as fse from 'fs-extra';
+import * as path from 'path';
+import * as vscode from 'vscode';
+import { IActionContext, IAzureQuickPickItem, } from "vscode-azureextensionui";
+import { ext } from "../extensionVariables";
+import { captureCancelStep } from '../utils/captureCancelStep';
+import { Platform, PlatformOS } from "../utils/platform";
+import { quickPickWorkspaceFolder } from '../utils/quickPickWorkspaceFolder';
+import { ConfigureTelemetryCancelStep, ConfigureTelemetryProperties, promptForPorts as promptForPortsUtil, quickPickOS } from './configUtils';
+
+export interface ScaffoldContext extends IActionContext {
+    folder?: vscode.WorkspaceFolder;
+    initializeForDebugging?: boolean;
+    os?: PlatformOS;
+    outputFolder?: string;
+    platform?: Platform;
+    ports?: number[];
+    rootFolder?: string;
+}
+
+export interface ScaffolderContext extends ScaffoldContext {
+    captureStep<TReturn, TPrompt extends (...args: []) => Promise<TReturn>>(step: ConfigureTelemetryCancelStep, prompt: TPrompt): TPrompt;
+    folder: vscode.WorkspaceFolder;
+    initializeForDebugging: boolean;
+    platform: Platform;
+    promptForOS(): Promise<PlatformOS>;
+    promptForPorts(defaultPorts?: number[]): Promise<number[]>;
+    rootFolder: string;
+}
```

Comment:
@philliphoff for my education, can you explain what is the intention of defining `ScaffolderContext` interface that extends `ScaffoldContext`, but most of the members of the new interface seem to re-use the names of the parent interface members, just making them non-optional? Should I think about this as "the new interface members override the old interface members"?

Except down below (starting at line 118) you are calling the scaffolder by constructing an object that seems to de-structure the context object (of type `ScaffoldContext`), then adds a bunch of properties, some named, and some just by position?? How can you be sure that the data you are passing is going to end up populating the right properties of the resulting `ScaffolderContext` object?

It has been a while since I was looking at TS code that uses advanced typing mechanisms and it is quite confusing to me

Reply:
`scaffold()` can be invoked either via a user-executed command or programmatically (e.g. by the IoT extension).  In the latter case, in particular, we want to be lenient in what's specified, filling in defaults where it makes sense.  Hence, the `ScaffoldContext` input to `scaffold()` makes most properties optional.  However, we want the individual scaffolders (e.g. for .NET, Node.js, etc.) to be able to depend on *some* but not *every* one of those values being specified (or, rather, we don't want each scaffolder to have its own custom logic for determining the defaults).  So `ScaffolderContext` inherits from `ScaffoldContext`, to pass along the optional properties, but then "overrides" those that will have been provided common defaults.  (In C#, you do something equivalent by using the `new` keyword in the interface/class.)

In line 118, we construct a new `ScaffolderContext` by copying all properties of the existing `ScaffoldContext` (`...context,`), setting some of the properties that only exist on `ScaffolderContext` (e.g. `promptForOS`), and then overrides the properties optional in `ScaffoldContext` but no longer optional in `ScaffolderContext`.  Note that the copy/override is done by property name and never by position, apart from copying/overriding occurring in left-to-right order.  Also note that we create a new context rather than update the original `ScaffoldContext` context in the interests of not modifying an object we don't own, and we copy values from the original `ScaffoldContext` context to the new object in the interests of passing along properties to the individual scaffolder that we may not explicitly know about (e.g. an additional property to configure the Node.js scaffolder), but that a caller may know about.

Reply:
Thanks Phil, this makes much more sense now.

Perhaps it is worth considering adding a comment that describes the intention between `ScaffoldContext` and `ScaffolderContex` (what you said above).

The last missing piece for me is the fact that, in Typescript object literals, `prop` is a shorthand for `prop: prop`, see https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#4.5

Reply:
Added some more comments.